source_mesh.py

```diff
@@ -0,0 +1,381 @@
+            confinement space (rad).
+        mbc_verts (list of EntityHandle): list of mesh vertices.
+        verts (list of list of float): list of 3D Cartesian coordinates of each
+            vertex in form [x (cm), y (cm), z (cm)].
+        verts_s (list of float): list of closed flux surface indices for each
+            vertex.
+            
+    Returns:
+        strengths (list of float): list of source strengths for each
+            tetrahedron (1/s).
+    """
+    # Instantiate tetrahedra mesh set in PyMOAB
+    tet_set = mbc.create_meshset()
+    # Add vertices to mesh set
+    mbc.add_entity(tet_set, mbc_verts)
+
+    # Initialize list of tetrahedra source strengths
+    strengths = []
+    
+    # Create tetrahedra, looping through vertices
+    for i, phi in enumerate(phi_list[:-1]):
+        # Define index for magnetic axis at phi
+        ma_id = i*((num_s - 1)*num_theta + 1)
+        # Define index for magnetic axis at next phi
+        next_ma_id = ma_id + (num_s - 1)*num_theta + 1
+
+        # Create tetrahedra for wedges at center of plasma
+        for k, theta in enumerate(theta_list[:-1], 1):
+            # Define indices for wedges at center of plasma
+            wedge_id0 = ma_id
+            wedge_id1 = ma_id + k
+            wedge_id2 = ma_id + k + 1
+            wedge_id3 = next_ma_id
+            wedge_id4 = next_ma_id + k
+            wedge_id5 = next_ma_id + k + 1
+
+            # Define three tetrahedra for wedge
+            create_tet(
+                mbc, tag_handle, tet_set, mbc_verts, verts, verts_s, wedge_id1,
+                wedge_id2, wedge_id4, wedge_id0, strengths
+            )
+            create_tet(
+                mbc, tag_handle, tet_set, mbc_verts, verts, verts_s, wedge_id5,
+                wedge_id4, wedge_id2, wedge_id3, strengths
+            )
+            create_tet(
+                mbc, tag_handle, tet_set, mbc_verts, verts, verts_s, wedge_id0,
+                wedge_id2, wedge_id4, wedge_id3, strengths
+            )
+
+        # Create tetrahedra for hexahedra beyond center of plasma
+        for j, s in enumerate(s_list[1:-1]):
+            # Define index at current closed magnetic flux surface
+            s_offset = j*num_theta
+            # Define index at next closed magnetic flux surface
+            next_s_offset = s_offset + num_theta
+
+            # Create tetrahedra for current hexahedron
+            for k, theta in enumerate(theta_list[:-1], 1):
+                # Define indices for hexahedron beyond center of plasma
+                tet_id0 = ma_id + s_offset + k
+                tet_id1 = ma_id + next_s_offset + k
+                tet_id2 = ma_id + next_s_offset + k + 1
+                tet_id3 = ma_id + s_offset + k + 1
+                tet_id4 = next_ma_id + s_offset + k
+                tet_id5 = next_ma_id + next_s_offset + k
+                tet_id6 = next_ma_id + next_s_offset + k + 1
+                tet_id7 = next_ma_id + s_offset + k + 1
```

Comment:
play with `numpy.ravel_multi_index()` to make this more readable by future you (or future me)